components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java

```diff
@@ -228,89 +230,92 @@ public OAuth2AccessTokenRespDTO issueAccessToken(OAuth2AccessTokenReqDTO tokenRe
+
                 if (refreshTokenFirst) {
-                    RefreshTokenValidationDataDO refreshTokenDO = tokenMgtDAO.validateRefreshToken(
-                            revokeRequestDTO.getConsumerKey(), revokeRequestDTO.getToken());
-                    if (refreshTokenDO != null && StringUtils.isNotEmpty(refreshTokenDO.getRefreshTokenState()) &&
-                            (OAuthConstants.TokenStates.TOKEN_STATE_ACTIVE.equals(
-                                    refreshTokenDO.getRefreshTokenState()) ||
-                                    OAuthConstants.TokenStates.TOKEN_STATE_EXPIRED.equals(
-                                            refreshTokenDO.getRefreshTokenState()))) {
-                        org.wso2.carbon.identity.oauth.OAuthUtil.clearOAuthCache(
-                                revokeRequestDTO.getConsumerKey(), refreshTokenDO.getAuthorizedUser(),
-                                OAuth2Util.buildScopeString(refreshTokenDO.getScope()));
-                        org.wso2.carbon.identity.oauth.OAuthUtil.clearOAuthCache(
-                                revokeRequestDTO.getConsumerKey(), refreshTokenDO.getAuthorizedUser());
-                        org.wso2.carbon.identity.oauth.OAuthUtil.clearOAuthCache(refreshTokenDO.getAccessToken());
-                        tokenMgtDAO.revokeTokens(new String[]{refreshTokenDO.getAccessToken()});
-                        addRevokeResponseHeaders(revokeResponseDTO, refreshTokenDO.getAccessToken(),
-                                revokeRequestDTO.getToken(), refreshTokenDO.getAuthorizedUser());
-                    } else {
-                        AccessTokenDO accessTokenDO = tokenMgtDAO.retrieveAccessToken(
-                                revokeRequestDTO.getToken(), true);
-                        if (accessTokenDO != null) {
-                            org.wso2.carbon.identity.oauth.OAuthUtil.clearOAuthCache(
-                                    revokeRequestDTO.getConsumerKey(), accessTokenDO.getAuthzUser(),
-                                    OAuth2Util.buildScopeString(accessTokenDO.getScope()));
-                            org.wso2.carbon.identity.oauth.OAuthUtil.clearOAuthCache(
-                                    revokeRequestDTO.getConsumerKey(), accessTokenDO.getAuthzUser());
-                            org.wso2.carbon.identity.oauth.OAuthUtil.clearOAuthCache(revokeRequestDTO.getToken());
-                            tokenMgtDAO.revokeTokens(new String[]{revokeRequestDTO.getToken()});
-                            addRevokeResponseHeaders(revokeResponseDTO, accessTokenDO.getAccessToken(),
-                                    accessTokenDO.getRefreshToken(), refreshTokenDO.getAuthorizedUser());
-                        }
+                    refreshTokenDO = tokenMgtDAO
+                            .validateRefreshToken(revokeRequestDTO.getConsumerKey(), revokeRequestDTO.getToken());
+                    if (refreshTokenDO == null) {
```

Comment:
Seems checking for null only isn't enough. In lines 251 to 255 of the original code we are doing much more conditional checks. Specifically we are not only checking if a token exists but we are also checking its status.